definition_validator/variance.cc

```diff
@@ -0,0 +1,184 @@
+
+#include "common/typecase.h"
+#include "core/Symbols.h"
+#include "core/core.h"
+#include "core/errors/resolver.h"
+
+#include "variance.h"
+
+using namespace std;
+
+namespace sorbet::definition_validator::variance {
+
+class VarianceValidator {
+private:
+    const core::Loc loc;
+
+    VarianceValidator(const core::Loc loc) : loc(loc) {}
+
+    core::Variance invertVariance(const core::Variance variance) {
+        switch (variance) {
+            case core::Variance::CoVariant:
+                return core::Variance::ContraVariant;
+            case core::Variance::Invariant:
+                return core::Variance::Invariant;
+            case core::Variance::ContraVariant:
+                return core::Variance::CoVariant;
+        }
+    }
+
+    string showVariance(const core::Variance variance) {
+        switch (variance) {
+            case core::Variance::CoVariant:
+                return ":out";
+            case core::Variance::Invariant:
+                return ":invariant";
+            case core::Variance::ContraVariant:
+                return ":in";
+        }
+    }
+
+    bool compatibleVariance(const core::Variance left, const core::Variance right) {
+        switch (left) {
+            case core::Variance::CoVariant:
+                return right != core::Variance::ContraVariant;
+            case core::Variance::Invariant:
+                return right == core::Variance::Invariant;
+            case core::Variance::ContraVariant:
+                return right != core::Variance::CoVariant;
+        }
+    }
+
+    void validate(const core::Context ctx, const core::Variance variance, const core::TypePtr type) {
+        typecase(
+            type.get(), [&](core::ClassType *klass) {},
+
+            [&](core::LiteralType *lit) {},
+
+            [&](core::SelfType *self) {},
+
+            [&](core::SelfTypeParam *sp) {},
```

Comment:
Is this ever encountered?

Reply:
Yes. I'll try to dig up the tests that needed this.

Reply:
`test/testdata/infer/self_type.rb` and `test/testdata/resolver/self.rb` both failed when these cases were removed.